Add spec of dotnet nuget config command

heng-liu opened the pull request:
Fixes: https://github.com/NuGet/Home/issues/12297
Add a spec of `dotnet nuget config list` command.

It's a part of epic issue:
 Add a dotnet nuget config command https://github.com/NuGet/Home/issues/12296

heng-liu committed:
work

heng-liu committed:
fix

erdembayar commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,61 @@
+The `dotnet nuget config list` is a community ask. We will consider adding more commands, like add/update/delete, in the future.
+
+## Open Questions
+1. Shall we add `working directory` as an option? So that people could check different NuGet configuration lists without changing current directory.
```
I believe this one should be in current spec. It's not hard to implement and just 1 option. For example, in CI pipeline logging it's useful to just pass working directory rather than switching to different working directories.

heng-liu replied:
Thanks for reviewing! I agree, this should be useful. Added this option :)

heng-liu committed:
address feedback

heng-liu committed:
fix

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
````diff
@@ -0,0 +1,74 @@
+
+- List all the NuGet configuration file that will be applied, when invoking NuGet command in the specific directory.
+```
+dotnet nuget config list --current-directory c:\repos\Solution
````
Do we need a `--current-directory` option. Can the location be the argument to the command and be the working directory if not specified.

erdembayar replied:
Yes, that's better. Just pass as argument instead of typing long option keyword.

heng-liu replied:
Great suggestion! Changed it into an argument.

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,74 @@
+### `dotnet nuget config`
+
+#### Commands
+If no command is specified, the command will default to list.
```
What happens today with commands such as sources and packages? 
Is there a default command?

heng-liu replied:
Good question! There is no default command for sources and packages, not sure if it's because verbs need to go first.
But there is a default command `list` for [dotnet nuget trust command](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-nuget-trust#commands).
So, when no sub command is specified, do we want to let people know all the subcommands for `dotnet nuget config`? Or just run dotnet nuget config list command by default?

heng-liu replied:
Hi @JonDouglas, may I know if you have any suggestions about this? Thanks! :)

nkolev92 replied:
Personally, I think we should align to dotnet list package.

heng-liu replied:
Changed to no default command. Thanks!

nkolev92 reviewed (commented):
I'm really excited about this addition :)

zivkan commented on proposed/2022/dotnet-nuget-config-spec.md:
````diff
@@ -0,0 +1,74 @@
+```
+dotnet nuget config list
+
+c:\repos\Solution\Project\NuGet.Config
````
We discourage customers trying to nuget.config files in project directories, because VS will never use it, and I'm not sure that command line solution restore will either. Therefore, we shouldn't use discourages practices in our examples.

heng-liu replied:
Thanks for letting me know! Changed the examples accordingly.

heng-liu committed:
address feedback

heng-liu committed:
address feedback

heng-liu commented:
There are two types of implementations:
1. Treat `list` as subcommand.  Example: [dotnet nuget trust](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-nuget-trust)
2. Treat `list` as an option. Example: [dotnet nuget locals](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-nuget-locals)

I followed the 1st in this spec. But I'm not quite sure if there is any preference. Please advise, thanks!

nkolev92 commented:
I think it should be a subcommand. 

The subcommand matches `dotnet list package`, `dotnet list reference` etc. 
It matches the direction of the dotnet CLI.

heng-liu commented:
Hi @baronfel, thanks for your help offline on the guidance!
Could you please take a look and let me know if there is anything I need to change? Thanks! :)

dominoFire commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,79 @@
+1. Do we need to add this into NuGet.exe CLI?
+
+## Considerations
+1. NuGet.exe [config command](https://learn.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-config) is implemented. But there is no `list` command. And the behavior is confusing (the `set` command will set the property which appears last when loading, so sometimes it's not updating the closest NuGet configuration file). Do we want to implement those subcommand(e.g.`set`) in the future i ...
```
I would say yes to this question. Let's create an issue for that.

nkolev92 replied:
> And the behavior is confusing (the `set` command will set the property which appears last when loading, so sometimes it's not updating the closest NuGet configuration file). Do we want to implement those subcommand(e.g.`set`) in the future in dotnet.exe differently?

I think that's a discussion for a future point. 

Personally, I'd prefer we guide users towards choosing the right config file instead of NuGet making assumptions. 

I'd be hesitant to have VS and NuGet.exe work differently from dotnet.exe.

heng-liu replied:
Thanks for reviewing! Moved it to "Future Work" section as we won't address it in the implementation.

zivkan commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,79 @@
+## Non-Goals
+Design and implement `dotnet nuget config` command with commands other than `list`, E.g. add/update/delete
+
+## Solution
+The following command will be implemented in the `dotnet.exe` CLI.
```
I really wish GitHub had threaded/resolvable conversations for comments without a specific file/line number 😔

I'd like to know what the design is for error scenarios, specifically when one or more of the XML files contains syntax errors (from an implementation point of view, this means that `Settings.Load*` apis will fail, so we can't use `loadedConfig.GetConfigFilePaths()`).

Is `dotnet nuget config list` going to report an error and fail to list all the files?  Will it at least tell us the filename of the XML that couldn't be parsed?  Or will all the XML files be listed despite the parsing failure?  Will it tell us which file(s) contain parsing errors?

heng-liu replied:
That's a good question!
I tested with NuGet.exe and dotnet.exe, and both of them are able to tell us the filename of the XML that couldn't be parsed, as below:
```
nuget list source
NuGet.Config is not valid XML. Path: 'C:\Users\username\Source\Repos\NuGet.Config'.
  The 'disabledPackageSources' start tag on line 19 position 4 does not match the end tag of 'configuration'. Line 20, position 3.
```
```
dotnet nuget list source
error: NuGet.Config is not valid XML. Path: 'C:\Users\username\Source\Repos\NuGet.Config'.
error:   The 'disabledPackageSources' start tag on line 19 position 4 does not match the end tag of 'configuration'. Line 20, position 3.
```
So I think the `dotnet nuget config list` is able to do this as well.

If we'd like to list all config files despite the parsing failure, we'd consider if it worth doing.
Since running all nuget commands will encounter this invalid XML error, users need to fix this invalid XML before running any NuGet commands. So this seems to be a prerequisite and there is no harm if user needs to do this first before knowing the list of config files.
In addition, this error message clearly points out the file path and location,  so it should be easy for users to correct the invalid XML.

The only scenario needs to be considered is, with the invalid XML error, user should able to specifying a specific correct config file to run nuget commands. And if we list all config file locations despite the parsing failure, it might help people to find the path easier. But it seems "listing all config file locations despite the parsing failure" adds very limited value.

Those are my preliminary thoughts. Please let me know if you have any other thoughts :)

The other error scenario I'd like to talk about is, the specified working directory doesn't exist. Shall we go ahead listing user-wide and machine-wide config files? Or show a warning saying the working directory doesn't exist? Or both?

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,79 @@
+The `dotnet nuget config list` is a community ask. We will consider adding more commands, like add/update/delete, in the future.
+
+## Open Questions
+1. Do we need to add this into NuGet.exe CLI?
```
IMO we do dotnet.exe first and have nuget.exe be a future consideration.

heng-liu replied:
Thanks! Moved it to "Future Work" section as we won't address it in the implementation.

heng-liu committed:
address comments

heng-liu commented:
Hi @baronfel, I think I've addressed all comments for now. 
May I know if you could take a look? Or do you prefer a review meeting late this week or early next week? Thanks!

erdembayar commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,84 @@
+3. NuGet.exe [config command](https://learn.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-config) is implemented. But there is no `list` command. And the behavior is confusing (the `set` command will set the property which appears last when loading, so sometimes it's not updating the closest NuGet configuration file). Do we want to implement those subcommand(e.g.`set`) in the future i ...
+
+## Open Questions
+1. When the specified `CURRENT_DIRECTORY` doesn't exist, shall we list user-wide and machine-wide config files? Or show a warning saying the working directory doesn't exist? Or do both?
```
I believe it just should error out/warn. If there is error, I'm not sure how much machine wide config matters.

nkolev92 replied:
I think we should just error too.

heng-liu replied:
Thanks! Updated with only showing a warning/error.

heng-liu committed:
address feedback

zivkan commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,84 @@
+## Open Questions
+1. When the specified `CURRENT_DIRECTORY` doesn't exist, shall we list user-wide and machine-wide config files? Or show a warning saying the working directory doesn't exist? Or do both?
+
+We decided to show a warning/error. 
```
This means it's no longer an open question, and the outcome can be integrated into the design above.

heng-liu replied:
You're right :) Integrated into the design and add an example about this scenario.

heng-liu commented:
I just thought about another scenario, when there is an error when accessing the current directories/or it's parent directories.
For now, nuget command just ignore the inaccessible config file and there will be no warning/error. As we check if the config file exists at this [line](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs#L823), and according to [doc](https://learn.microsoft.com/en-us/dotnet/api/system.io.file.exists?view=net-7.0):
`The Exists method returns false if any error occurs while trying to determine if the specified file exists. This can occur in situations that raise exceptions such as passing a file name with invalid characters or too many characters, a failing or missing disk, or if the caller does not have permission to read the file.`

So, for a folder as below, if I change "Solution" folder to inaccessible, the "NuGet.Config" file under this folder will be ignored and there will be no warning/error.

C:\TEST
│   NuGet.Config
│
└───Repos
    │   NuGet.Config
    │
    └───**Solution**
        │   **NuGet.Config**
        │   Solution.sln
        │
        └───Project
                Class1.cs
                Project.csproj

I'm assuming this is expected (Please let me know if you consider this is a bug). Shall I warn/error for this situation for "dotnet nuget config list" command? Or just ignore it and show the rest of NuGet.Config files without any warning/error(align with other NuGet commands)?

kartheekp-ms commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,84 @@
+
+#### Arguments
+
+- CURRENT_DIRECTORY
```
Incase if we are referring to `CURRENT_DIRECTORY` as working directory, `dotnet store` command has `-w|--working-dir <WORKING_DIRECTORY>` option https://learn.microsoft.com/dotnet/core/tools/dotnet-store#optional-options. I think it is better to use the same option name for this command also.

heng-liu replied:
Thanks for reviewing! Renamed it as `WORKING_DIRECTORY`, but still an argument based on previous discussions. Pls let me know if you have any different thoughts :)

heng-liu committed:
address comments & add more examples

nkolev92 commented:
> I'm assuming this is expected (Please let me know if you consider this is a bug). Shall I warn/error for this situation for "dotnet nuget config list" command? Or just ignore it and show the rest of NuGet.Config files without any warning/error(align with other NuGet commands)?

I'm guessing we match what commands that use the configs do, for example running restore.

heng-liu commented:
> I'm guessing we match what commands that use the configs do, for example running restore.

You're right. It matched what commands that use the configs.

heng-liu commented:
Hi @baronfel , may I know if you have some time to take a look at this PR this week? Thanks!

kartheekp-ms commented:
https://github.com/NuGet/Home/issues/12213 issue may be out of scope for this spec, but the feature request make sense to me.

`When I issue a nuget config command in a repository, I don't get any information. And if I specify -Verbosity detailed, then I only get the version number of the nuget.exe executed.
It would be nice to have nuget config list the current configuration values as being evaluated from the various config files, and if -Verbosity detailed specified also the config file being source for given config value.`

bstordrup commented:
> #12213 issue may be out of scope for this spec, but the feature request make sense to me.
> 
> `When I issue a nuget config command in a repository, I don't get any information. And if I specify -Verbosity detailed, then I only get the version number of the nuget.exe executed. It would be nice to have nuget config list the current configuration values as being evaluated from the various config files, and if -Verbosity detailed specified also the config file being source for given config value.`

Regarding the #12213, I have a local branch that implements this by adding a `all` dummy item and is used like this:

```
nuget config all -Verbosity detailed
```

To have the -Verbosity work correctly however, the Origin property (and hence SettingsFile class) need to be added to PublicAPI, but these changes are also done in my local branch.

Once this pull request is merged, I can update my branch and make a pull request for the change in #12213

heng-liu committed:
add verbosity detailed

heng-liu commented:
Thanks @bstordrup and @kartheekp-ms ! I've added the `-verbosity` option into this spec. And the merged NuGet configuration will be displayed when verbosity is detailed.

zivkan commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,146 @@
+- -v|--verbosity <LEVEL>
+
+Sets the verbosity level of the command. Allowed values are q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic]. The default is minimal. 
+When the verbosity level is detailed or diagnostic, the command will display the merged NuGet configuration.
```
It feels out of place to me to implement "display merged config" as the detailed output for "Lists all the NuGet configuration file locations". 

I feel like the functionality belongs in a `get` verb, not `list`.

I have no idea if anyone will actually want to do it, but also imagine that someone wants to have a script where the output is piped to a file, and then they use an XML processing library/tool to work on the XML.  With this design, it's difficult because of all the non-XML output. Whereas a dedicated command to get the merged config could have only XML output at minimal verbosity, allowing customers to "safely" process the XML output of something like `dotnet nuget config get all -v m`.

You might also want to have a look at `git config` and all its options as prior-art that customers might already be familiar with. Admittedly they use `--list` to list all the config values (what you're calling the "merged config"), rather than listing all the files that contributed. I don't see an option to only list gitconfig files used, but they do have a `--show-origin`, so `git config --list --show-origin` will show all the config key-value pairs, and which

nkolev92 replied:
I agree, verbosity should only really be a thing when you want to know more about certain config value. 

IMO here's how it should work longer term: 

- dotnet nuget list config-> Lists all NuGet config file paths.
- dotnet nuget list config globalPackagesFolder -> Gives you the value of the globalPackagesFolder configuration. 
- dotnet nuget add config globalPackagesFolder D:\packages -> Sets the global packages folder in the approrpiate config. (same thing would apply to update)
- dotnet nuget list config globalPackagesFolder --verbosity detailed -> Gives you the value of the global packages folder configuration + the config where it comes from.

I know nuget.exe uses `get` and `set`, but the dotnet CLI uses add/remove/update so we should think in those verbs. 

Note the ordering to match `dotnet add package` and the like commands.

nkolev92 replied:
Apologies for not noticing the ordering and being on autopilot. 

dotnet commands are verb driven.

heng-liu replied:
Thanks @zivkan for the great suggestion!
I'm thinking if the following sounds good:
`dotnet nuget config list` to list all config values(merged configurations)
`dotnet nuget config list --show-path` to list all config values(merged configurations) and the config file paths.

//cc @JonDouglas

heng-liu replied:
> Apologies for not noticing the ordering and being on autopilot.
> 
> dotnet commands are verb driven.

Thanks @nkolev92 ! I asked Chet about dotnet command conventions and he helped me confirm that: `At this time we have semi-standardized on dotnet NOUN VERB, so dotnet nuget config list would be the preferred form for this command.` (In email: Re: Please advise on naming of a new dotnet nuget command)

nkolev92 replied:
my bad if I'm providing conflicting feedback :D 

Ignore the ordering then. I'm curious what you think about the other behavior.

heng-liu replied:
Thanks @nkolev92 ! Most of them sounds good to me. But seems we still don't have an option to view all config values (what customer mentioned)
So how about the following:

- `dotnet nuget config list`    list all config values (key + value).
- `dotnet nuget config list` --verbosity detailed   list all config values (key + value), and the config file paths at the bottom.
- `dotnet nuget config list <configKey>`  list the value of the specified config key
- `dotnet nuget config list <configKey> --verbosity detailed`    list the value of the specified config key, and the config file path it comes from
-  `dotnet nuget config add <configKey> <configValue> <config file path>`  add the key value into the specified config file
-  `dotnet nuget config update <configKey> <configValue> <config file path>`  update the key value pair in the specified config file
-  `dotnet nuget config delete <configKey> <config file path>` delete the key value pair from the specified config file

But I'm not sure what shall we do if the configuration is something with child attributes, but not a simple key-value pair. E.g. package source, packageSourceMapping etc. Shall we avoid providing add/update/delete command for those configurations?

zivkan replied:
Should we consider `set` and `unset` instead of `add`, `update`, and `delete`?  That's one less verb, and I don't think that `add` vs `update` matters to customers. (for example, would we want `add` to fail if it already exists in the config?   Would we want to fail `update` if the key doesn't exist in the config file?)

> `dotnet nuget config list <configKey>` list the value of the specified config key

nitpick/very opinionated: A single item is technically a special case of a list, but 

> But I'm not sure what shall we do if the configuration is something with child attributes,

nuget.exe config only handles the [`<configuration><config>` part of the config](https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file#config-section), which are key-value pairs (probably implemented by `Dictionary<string, string>`). The `nuget.exe config` commands could never be used for other parts of the config file, like sources, credentials, package source mapping, etc.  I'm not saying this is good, but be aware that if you want to make `dotnet nuget config`'s scope so much different to `nuget.exe config`'s existing implementation, it's going to make implementing the spec a lot more difficult.  It's definitely not ideal to design features based on how easy it is to implement, but sometimes we need to be pragmatic. As you pointed out, what would `dotnet nuget config add` commands even look like for package sources, credentials, or package source mapping?

nkolev92 replied:
> Most of them sounds good to me. But seems we still don't have an option to view all config values (what customer mentioned)

I think we might need a magic word for list all configs such as "all", something we promise to never make a config value for example. 

> Should we consider set and unset instead of add, update, and delete? That's one less verb, and I don't think that add vs update matters to customers. (for example, would we want add to fail if it already exists in the config? Would we want to fail update if the key doesn't exist in the config file?)

I think we should use the same verb as the CLI.

IMO, it's ok with config doesn't handle sources, credentials or source mapping. All of those will have their own individual commands.

zivkan replied:
FWIW, here's a customer complaint that separate add and update verbs are annoying: 
* https://github.com/dotnet/sdk/issues/29026#issue-1445497084

> I think we should use the same verb as the CLI.

Consistency is good, unless it's consistently user unfriendly 😁

erdembayar commented:
@heng-liu 
Maybe it's off topic. But could this detect nuget.config issues like broken xml tag or duplicate keys etc? It's hard for customers to detect exactly which entry key is causing problem. For example: https://developercommunity.visualstudio.com/t/After-upgrading-Visual-Studio-2022-to-17/10199374#T-ND10201096

heng-liu commented:
> @heng-liu Maybe it's off topic. But could this detect nuget.config issues like broken xml tag or duplicate keys etc? It's hard for customers to detect exactly which entry key is causing problem. For example: https://developercommunity.visualstudio.com/t/After-upgrading-Visual-Studio-2022-to-17/10199374#T-ND10201096

Thanks for reviewing! I don't think the command will be able to detect broken xml. Please see comment at https://github.com/NuGet/Home/pull/12172#discussion_r1005260706
For duplicate keys, I'm not sure if this is the same with [this one](https://github.com/NuGet/Home/issues/8407#issuecomment-1245890475). If so, it seems an bug not related to this spec. 
So generally speaking, for incorrect settings, I think we should have better error messages.

heng-liu committed:
address feedback, list merged settings by default

heng-liu committed:
add set and unset

heng-liu committed:
fix and add a question

heng-liu committed:
add consideration

heng-liu committed:
fix format

kartheekp-ms commented on proposed/2022/dotnet-nuget-config-spec.md:
````diff
@@ -0,0 +1,298 @@
+```
+dotnet nuget config list --verbosity detailed
+
+<configuration>
````
customers may not be interested in the XML content. How about just printing key/value pairs from the merged NuGet.Config file?

nkolev92 replied:
Personally I don't think this is valuable in V0, but I won't oppose it as long as others are good with that :)

heng-liu replied:
Changed into non-XML format. Thanks!

baronfel commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,298 @@
+
+This command will list merged NuGet configuration settings from one or multiple NuGet configuration files that will be applied, when invoking NuGet command from the current working directory path. 
+
+#### Arguments
+
+- WORKING_DIRECTORY
+
+Run this command as if working directory is set to the specified directory.
+
+If the specified `WORKING_DIRECTORY` doesn't exist, an error is displayed indicating the `WORKING_DIRECTORY` doesn't exist.
+
+> [!Note]
+> If `WORKING_DIRECTORY` (or its parent directories) is not accessible, the command will ignore any NuGet configuration files under those directories without any warning/error. This is aligned with other NuGet commands.
```
This argument feels odd to me - it's not used in any of the other proposed commands (instead explicit `--config-file` options are provided), so why does this concept get first-class treatment here?

Looking at the other commands (set, unset) to me `list` implies listing configuration values (parallel to setting/unsetting them), _not_ locating and showing the entire xml config file.  This command as described feels like a `dotnet nuget config paths` kind of command as a result.

baronfel replied:
The verb `get` is missing here, and it seems like a natural analogue for users that are scripting. Are scripting use cases a motivation here?

kartheekp-ms replied:
`dotnet store` command has `-w|--working-dir <WORKING_DIRECTORY>` option https://learn.microsoft.com/dotnet/core/tools/dotnet-store#optional-options

heng-liu replied:
Thanks for reviewing!
The option is renamed to `-w|--working-dir <WORKING_DIRECTORY>`. Thanks @kartheekp-ms!
The `get` is added. Thanks @baronfel !

nkolev92 replied:
> Thanks for reviewing! The option is renamed to `-w|--working-dir <WORKING_DIRECTORY>`. Thanks @kartheekp-ms! The `get` is added. Thanks @baronfel !

I don't think that was the outcome in the meeting. 

When we brought this up for discussion, Daniel pointed out that other commands such as `dotnet build`, `dotnet nuget sources` etc all use a directory or a project file as the first argument. 

I was under the impression that no one disagreed with that? 

I guess there's a recording so we can always figure it out from there :)

baronfel reviewed (commented):
Brief comment on list especially.  I have a couple follow up questions on the user experience though - is tab completion a thought here? Specifically it seems like set/unset are prime opportunities to suggest config keys that are valid.

heng-liu committed:
work

heng-liu committed:
address comments

heng-liu committed:
edit the open questions

heng-liu committed:
fix

heng-liu committed:
fix format

heng-liu committed:
fix

heng-liu committed:
fix option working-dir

heng-liu commented:
The spec is updated based on our last review meeting.
The main changes are:
dotnet nuget config path : show list of paths of config files
dotnet nuget config get : get all merged configuration settings (XML format)
dotnet nuget config get -v d : get all merged configuration settings with source of config file path (XML format)
dotnet nuget config get <CONFIG_KEY> : get the value of the configuration setting (a string, not XML format)
dotnet nuget config set <CONFIG_KEY> <CONFIG_VALUE>: set the configuration setting
dotnet nuget config unset <CONFIG_KEY> : unset the value of the configuration setting
Please take another look, thank you!

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,358 @@
+
+#### Commands
+
+- Path
```
Should the config be `paths` instead? 
I'm not really sure what the guidance for plural vs singular is.

heng-liu replied:
Thanks for reviewing! I used `path` in the spec and add this question into open question as #1. 
Hi @baronfel, could you help advise? Thanks!

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,358 @@
+
+- Path
+
+List all the paths of NuGet configuration files that will be applied, when invoking NuGet command from the current working directory path.
```
Shouldn't the get command be listed here?

heng-liu replied:
Thanks! Changed the structures accordingly.

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,358 @@
+
+#### Options
+
+- -w|--working-dir <WORKING_DIRECTORY>
```
I left the feedback on the original thread. I don't think we want an option. A working directory argument is a natural solution.
`--working-dir` is redundant imo.

nkolev92 replied:
Replying to myself again, this isn't redundant but a way to differentiate between getting all and specifying a directory.

heng-liu replied:
Thanks! Yes, @kartheekp-ms helped find an example of this option:
The Dotnet store command has -w|--working-dir <WORKING_DIRECTORY> option https://learn.microsoft.com/dotnet/core/tools/dotnet-store#optional-options
Since it can be short for `-w`, I think it should be okey to be an option?

nkolev92 replied:
The existence of one instance doesn't make it the best approach imo. 

I'm also one of the few people even commenting on this concern, so I'm not really sure how I feel about it because of that. Whether others don't care, or others prefer the working directory approach.

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
````diff
@@ -0,0 +1,358 @@
+The configuration file under C:\Test\AccessibleRepos\NotAccessibleSolution\NuGet.Config will be ignored without any warning or error.
+
+```
+dotnet nuget config list  --working-dir C:\Test\AccessibleRepos\NotAccessibleSolution
````
You probably just missed changing the list argument here.

heng-liu replied:
Good catch! Thanks! Fixed.

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
````diff
@@ -0,0 +1,358 @@
+```
+
+#### Commands
+- Get
````
Ah, now I see it. You just have commands subheadings. I'd recommend not having the same headers in a markdown down as it breaks linking. 

I'd the following: 

`### Commands
#### path
#### set
#### set
```

Hope that helps.

heng-liu replied:
Fixed. Thanks!

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
````diff
@@ -0,0 +1,358 @@
+
+Key 'http_proxy' not found.
+
+```
````
I think we're missing a more natural/more valuable idea. 

`dotnet nuget config get http_proxy --verbosity detailed`  could be: 
http://company-squid:3128@contoso.com" - C:\Users\username\AppData\Roaming\NuGet\NuGet.Config

I think this is is way more valuable than the  `dotnet nuget config get` xml format.

heng-liu replied:
Added this as an example. Thanks!

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,358 @@
+## Open Questions
+1. To show configuration files locations, is it better to use `--verbosity` option or some option named like `--show-path`? (git command is using `--show-origin` for similar purpose)
+
+2. `dotnet nuget config get <CONFIG_KEY>` will get the value of the specifc config key (which is aligned with NuGet config command). But
```
I think it's ok if `dotnet nuget config get` shows all the config settings, I just don't think we should show the XML.

heng-liu replied:
Thanks! Changed the format accordingly. We might need to go through all configuration sections for the formats later.

nkolev92 commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,358 @@
+
+3. `dotnet nuget config get -v d` will display source of each configuration setting key, in a format of comment in xml file. So that people can still redirect the output into a file without breaking any syntax. Does that sound good to you?
+
+3. `--working-dir` is changed from an argument into an option. Because we could not differentiate `dotnet nuget config get <WORKING_DIRECTORY>` and `dotnet nuget config get <CONFIG_KEY>`. Any better ideas?
```
I see the motivation for this change. 

IMO we should add a magic word for getting all the config values, example `all`.

heng-liu replied:
I added `all` and also add this as #2 in the open questions.
Hi @baronfel , could you pls advise? Thanks!

baronfel replied:
Is "all" necessary if there's no longer a 'working dir' argument? In that case, no argument value would mean `all` because the presence of an argument would be a filter.

nkolev92 replied:
my suggestion is we do the magic word `all` and make the argument the working directory.

`--working-directory`  just seems off. 

I'm also very conscious of the fact that only a few of us have commented on this topic in here, while more did in the meeting :D

nkolev92 reviewed (commented):
This feature is turning out bigger than originally expected, but I think that's for the better! 
A couple of suggestion for tweaks.

Great job @heng-liu

heng-liu committed:
address feedback

heng-liu committed:
Reformat the output of dotnet nuget config all

heng-liu committed:
address feedback

heng-liu committed:
address feedback

nkolev92 commented:
Great summary with the open questions @heng-liu 
I know it's been a lot of iterations, but I really appreciate that the open questions are up to date :)

heng-liu committed:
change WORKING_DIRECTORY from option to argument

heng-liu commented:
Hi @baronfel and @dsplaisted , there are 3 open questions are mainly about dotnet conventions. Could you please advise? Thanks!
Other comments are all addressed. (You may find the details in email I sent on 12/6 ).
Please let me know if you have any other concerns. Thanks!

zivkan commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,366 @@
+
+- --configfile <FILE>
+
+The NuGet configuration file (nuget.config) to use. If specified, only the settings from this file will be used. If it's not specified, `%AppData%\NuGet\NuGet.Config` (Windows), or `~/.nuget/NuGet/NuGet.Config` or `~/.config/NuGet/NuGet.Config` (Mac/Linux) is used. See [On Mac/Linux, the user-level config file location varies by tooling.](https://learn.microsoft.com/en-us/nuget/consume-packages/c ...
```
So, even if there's a nuget.config in the current directory, if `--configfile` is not specified, the local directory file will not be modified?

While I think it's good that there's an easy way to modify the user-profile nuget.config, I feel like there will be more customers who expect the local directory file to be edited by default, so I think this will be non-intuitive.

heng-liu replied:
Yes. This is the behavior in NuGet.exe config command right now. 
You may refer to the [doc ](https://learn.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-config#options), or the [code](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs#L208-L230) .
The code indicates if there is no clear tag, default to furthest from the user, which is the user-wide configuration file (machine wide configuration files are excluded as they have `IsReadOnly` flag set to true, so they are not writable).
Unless this is not the expected behavior in NuGet.exe, otherwise we've discussed to keep the behavior consistent between NuGet.exe and dotnet.exe, and we only consider changing the behavior when major version changes, if there are enough upvotes. (Please refer to [Future Work  #2](https://github.com/NuGet/Home/blob/779cfc85bdff9887f47ae36d5705151667a80dcb/proposed/2022/dotnet-nuget-config-spec.md#future-work) in the spec).

dsplaisted commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,366 @@
+Considering this will be a breaking change, we will only consider doing it in main version change, if there are enough votes.
+
+## Open Questions
+1. Shall we use `dotnet nuget config path` or `dotnet nuget config paths` to get all configuration file paths?
```
I would go with the plural (`dotnet nuget config paths`) since it returns multiple paths.

heng-liu replied:
Thanks @dsplaisted for reviewing! Changed into `paths`.

dsplaisted commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,366 @@
+## Open Questions
+1. Shall we use `dotnet nuget config path` or `dotnet nuget config paths` to get all configuration file paths?
+
+2. `WORKING_DIRECTORY` is changed from an option to an argument. So we have `dotnet nuget config path <WORKING_DIRECTORY>` and `dotnet nuget config get <ALL|CONFIG_KEY> <WORKING_DIRECTORY>`. Is it okey if we have two arguments in the second command? And if it's key, only the second argument could be optional, right? 
```
This seems OK to me, but I'd like to hear from and defer to @baronfel, as he's the expert.

heng-liu replied:
Hi @baronfel , could you help with this open question? Thanks!

baronfel replied:
Good eye on the potential for ambiguity of the grammar here. Let me make sure I have the proposed grammar correct:

> dotnet nuget config get <all|CONFIG_KEY> [<WORKING_DIRECTORY>]

meaning 
> the first argument is required, and will either be the string 'all' or the name of an existing config key. the second key is optional and if present represents the working directory to locate the nuget config file from

Is that correct? if so, **I see no problem and am ok with the proposal**. in cases where there is some potential for ambiguity, consider adding a flag to allow for explicit disambiguation (i.e. `-w <path>` like we rejected earlier), but we've already had that conversation so I don't think we need to revisit it.

heng-liu replied:
Yep, that's correct :)
Great! Thanks for the confirmation! I moved this open question to Considerations.

dsplaisted commented on proposed/2022/dotnet-nuget-config-spec.md:
````diff
@@ -0,0 +1,366 @@
+- Get `http_proxy` from config section, when invoking NuGet command in the current directory. Show the source(nuget configuration file path) of this configuration setting.
+
+```
+dotnet nuget config get http_proxy 
````
Is there supposed to be a "show path" or verbosity parameter here?

heng-liu replied:
Good catch! Added `--show-path`. Thanks!

dsplaisted commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,366 @@
+
+2. `WORKING_DIRECTORY` is changed from an option to an argument. So we have `dotnet nuget config path <WORKING_DIRECTORY>` and `dotnet nuget config get <ALL|CONFIG_KEY> <WORKING_DIRECTORY>`. Is it okey if we have two arguments in the second command? And if it's key, only the second argument could be optional, right? 
+
+3. To show configuration files locations, is it better to use `--verbosity` option or some option named like `--show-path`? (git command is using `--show-origin` for similar purpose)
```
`--show-path` is better in my opinion.  I think of verbosity as something to apply to logging, not so much the output of a command like this.  Also in most cases of the .NET CLI, the verbosity doesn't do anything unless it's a command that ends up running MSBuild.

heng-liu replied:
Thanks for the explanation! Make sense! Changed into `--show-path`.

heng-liu committed:
address feedbacks

heng-liu committed:
fix a typo

heng-liu committed:
address feedback

heng-liu commented:
Thanks all for helping review this spec!
I know there are lots of iterations. Thanks a lot for your patience!

All comments are addressed in this spec. It's ready to review again.

erdembayar commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,369 @@
+
+### Commands
+
+### Paths
```
Sorry, I might be missing the context. Why don't default to this command? Why need `dotnet nuget config paths` instead of just `dotnet nuget config`?

heng-liu replied:
No worries! In our review meeting, we discussed and agreed that showing config file paths and showing config file contents should using different words. So we use `dotnet nuget config paths` for showing paths and `dotnet nuget config get` for showing contents.
And `dotnet nuget config` will show the options just like `dotnet nuget config -help`.

erdembayar commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,369 @@
+
+- WORKING_DIRECTORY
+
+Run this command as if working directory is set to the specified directory. If it's not specified, the current working directory will be used.
```
If can choose to not pass anything isn't this one is `option` rather than `argument`?

heng-liu replied:
Not exactly. Here is the definition of [arguments](https://learn.microsoft.com/en-us/dotnet/core/tools/#arguments) and [options](https://learn.microsoft.com/en-us/dotnet/core/tools/#options) in dotnet command world.  
An example of argument is a solution or a project. it can be optional as well.
The folder path is something similar with a solution or a project, so we choose to use it as an argument. We checked this with  .NET sdk team in the review meetings.

erdembayar commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,369 @@
+If the specified `WORKING_DIRECTORY` doesn't exist, an error is displayed indicating the `WORKING_DIRECTORY` doesn't exist.
+
+> [!Note]
+> If `WORKING_DIRECTORY` (or its parent directories) is not accessible, the command will ignore any NuGet configuration files under those directories without any warning/error. This is aligned with other NuGet commands.
```
Here not accessible means "path exist, but don't enough privilege/permission" right?

heng-liu replied:
Good question! Yes, it means the path exist but don't have enough privilege/permission to access.

erdembayar commented on proposed/2022/dotnet-nuget-config-spec.md:
```diff
@@ -0,0 +1,369 @@
+
+- CONFIG_VALUE
+
+Set the value of the CONFIG_KEY to CONFIG_VALUE.
```
If value type doesn't match with expected value type, then does it show appropriate message?
If you include 1 unhappy path scenario output for each use case, then that would be great. I learned this from `dotnet list package --format json` spec.

heng-liu replied:
Thanks @erdembayar ! That's a good suggestion!
However, after checking, NuGet.exe config -Set doesn't check the key at all. So if we run `NuGet.exe config -Set ABC=abc`, the following will be added into NuGet.Config file:
```
  <config>
    <add key="ABC" value="abc" />
  </config>
```
Created a follow-up issue https://github.com/NuGet/Home/issues/12345, added it to the epic https://github.com/NuGet/Home/issues/12296.

nkolev92 reviewed (approved):
Love it. 
Great job.

heng-liu commented:
I'll merge the PR since it's approved several weeks ago and there is no new comments. 
Please let me know if there is anything you'd like to discuss and I will create follow-up issues if necessary. Thanks!
FYI @dsplaisted @baronfel
